src/compiler/emitter.ts

```diff
@@ -477,6 +479,7 @@ export function forEachEmittedFile<T>(
 }
 
 export function getTsBuildInfoEmitOutputFilePath(options: CompilerOptions) {
+    if (some(internalOptionDeclarations, d => hasProperty(options, d.name))) return undefined; // disable buildinfo if an internal option is set

```

Comment:
i dont think we need this in PR at all. But i like the addition of internal options in commandline parser option declarations

Reply:
Why shouldn't they? `allowNonTsExtensions` affects both diagnostics and files loaded, both of which go into a `.buildinfo`. `suppressOutputPathsCheck` affects diagnostics. There's probably a slightly more appropriate invalidation level than "throw out the build info entirely" for both, but it's not like we've picked what those are, right? And we don't really need to, since they're `@internal` - they quite literally cannot be used with `--build` or even via APIs without exposing TS internals.

Reply:
if program has more files than before it is handled by incremental build correctly so you dont need to encode in buildInfo and hence this should not affect buildInfo. BuildInfo is not just for `--build` but normal tsc builds with incremetnal on as well. So even if someone wrote buildInfo with this option set the change in program will recalculate errors and do things correctly.

Reply:
You could end up with the same number, just different resolutions - `allowNonTsExtensions` can make us load a `"./foo"` over a `"./foo.js"` when referenced by a `"./foo"` specifier.

Reply:
Builder handles change in resolutions

Reply:
Er, but doesn't that imply it should really be `affectsModuleResolution` to get all the proper invalidation and caching behavior for swapping resolutions around?

Reply:
Right, but it doesn't, and hasn't, because internal options are ones we generally haven't invested in getting to work well with `--build` mode yet, and have yet to select and appropriate set of cache invalidation fields. And that's OK, since they're internal they're quite literally impossible to use with `--build` (even via API) without poking at our internals to expose the options. So we could take the approach that all internal options cause maximal invalidation, in lieu of properly setting up other settings, as a stop-gap. Or we could just do nothing for all of them (which is what we currently do), since they quite literally cannot be used without using non-public APIs, so how `--build` behves with them is rightfully undefined.